src/auth/src/credentials/external_account_sources/executable_sourced.rs

```diff
@@ -0,0 +1,317 @@
+                    None => Duration::from_secs(DEFAULT_TIMEOUT_SECS.into()),
+                };
+                Self::from_command(command, timeout).await
+            }
+            _ => Err(CredentialsError::from_msg(
+                false,
+                format!("{MSG}, either `output_file` or `command` needs to be informed"),
+            )),
+        }?;
+
+        let output = output.trim().to_string();
+        let subject_token = Self::parse_token(output)?;
+
+        if subject_token.is_empty() {
+            let msg = format!("{MSG}, subject token is empty");
+            return Err(CredentialsError::from_msg(false, msg));
+        }
+
+        Ok(subject_token)
+    }
+}
+
+impl ExecutableSourcedCredentials {
+    async fn from_output_file(output_file: String) -> Result<String> {
+        let content = std::fs::read_to_string(output_file)
+            .map_err(|e| CredentialsError::from_source(false, e))?;
+        Ok(content)
+    }
+
+    async fn from_command(command: String, timeout: Duration) -> Result<String> {
+        // For security reasons, we need our consumers to set this environment variable to allow executables to be run.
+        let allow_executable = std::env::var(ALLOW_EXECUTABLE_ENV)
+            .ok()
+            .unwrap_or("0".to_string());
+        if allow_executable != "1" {
+            return Err(CredentialsError::from_msg(
+                false,
+                "executables need to be explicitly allowed (set GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES to '1') to run",
+            ));
+        }
+
+        let (command, args) = Self::split_command(command);
+        let output = Command::new(command.clone()).args(&args).output();
+        let output = tokio_timeout(timeout, output.into_future())
+            .await
+            .map_err(|e| CredentialsError::from_source(true, e))?
+            .map_err(|e| CredentialsError::from_source(true, e))?;
+
+        if !output.status.success() {
+            let stderr = String::from_utf8(output.stderr)
+                .map_err(|e| CredentialsError::from_source(false, e))?;
+            let msg = format!("{MSG}, command execution failed, stderr=<{stderr}>");
+            if let Some(code) = output.status.code() {
+                let msg = format!("{msg}, code={code}");
+                return Err(CredentialsError::from_msg(true, msg));
+            };
+            return Err(CredentialsError::from_msg(true, msg));
+        }
+
+        let subject_token = String::from_utf8(output.stdout)
+            .map_err(|e| CredentialsError::from_source(true, e))?
+            .to_string();
+
+        Ok(subject_token)
+    }
+
+    fn split_command(command: String) -> (String, Vec<String>) {
+        let mut parts = command.split_whitespace();
+
+        if let Some(command) = parts.next() {
+            let args: Vec<String> = parts.map(String::from).collect();
+            return (command.to_string(), args);
+        }
+
+        (command, vec![])
```

Comment:
I think you can simplify this to:


```suggestion
    let mut parts = command.split_whitespace();
    let Some(cmd) = parts.next() else {
        return (command, vec![]);
    };
    (cmd.to_string(), parts.map(String::from).collect())
}
```

Reply:
Aside: this is a really terrible design in the input file format, they should have used a struct with the command and the arguments.  It is impossible to use arguments (or executable paths) that have spaces in them with the current design.

Reply:
TIL that inline `else`, thanks for the tip.